sdk/src/main/scala/io/horizen/certificatesubmitter/AbstractCertificateSubmitter.scala

```diff
@@ -269,33 +270,44 @@ abstract class AbstractCertificateSubmitter[
   }
 
   protected def calculateSignatures(messageToSign: Array[Byte], signersPublicKeys: Seq[SchnorrProposition]): Try[Seq[CertificateSignatureInfo]] = Try {
-    def getSignersPrivateKeys(sidechainNodeView: View): Seq[CertificateSignatureInfo] = {
+
+    def getSignersLocalPrivateKeys(sidechainNodeView: View): Seq[(SchnorrSecret, Int)] = {
       val wallet = sidechainNodeView.vault
-      val privateKeysWithIndexes = signersPublicKeys.map(signerPublicKey => wallet.secret(signerPublicKey)).zipWithIndex.filter(_._1.isDefined).map {
+      signersPublicKeys.map(signerPublicKey => wallet.secret(signerPublicKey)).zipWithIndex.filter(_._1.isDefined).map {
         case (secretOpt, idx) => (secretOpt.get.asInstanceOf[SchnorrSecret], idx)
       }
+   }
 
-      val remainingKeys = signersPublicKeys.zipWithIndex.filterNot(key_index => privateKeysWithIndexes.map(_._2).contains(key_index._2))
-      (signaturesFromEnclave(messageToSign, remainingKeys)
-        ++ privateKeysWithIndexes.map {
-        case (secret, pubKeyIndex) => CertificateSignatureInfo(pubKeyIndex, secret.sign(messageToSign))
-      })
-    }
+    val privateKeysWithIndexes = Await.result(sidechainNodeViewHolderRef ? GetDataFromCurrentView(getSignersLocalPrivateKeys), timeoutDuration)
+      .asInstanceOf[Seq[(SchnorrSecret, Int)]]
+    val remainingKeys = signersPublicKeys.zipWithIndex.filterNot(key_index => privateKeysWithIndexes.map(_._2).contains(key_index._2))
+    (signaturesFromEnclave(messageToSign, remainingKeys)
+      ++ privateKeysWithIndexes.map {
+      case (secret, pubKeyIndex) => CertificateSignatureInfo(pubKeyIndex, secret.sign(messageToSign))
+    })
 
-    Await.result(sidechainNodeViewHolderRef ? GetDataFromCurrentView(getSignersPrivateKeys), timeoutDuration)
-      .asInstanceOf[Seq[CertificateSignatureInfo]]
   }
 
   def signaturesFromEnclave(messageToSign: Array[Byte], indexedPublicKeys: Seq[(SchnorrProposition, Int)]): Seq[CertificateSignatureInfo] = {
     if (!secureEnclaveApiClient.isEnabled) return Seq()
-    val signaturesFromEnclaveFuture = secureEnclaveApiClient.listPublicKeys()
-      .map(managedKeys => indexedPublicKeys.filter(key_index => managedKeys.contains(key_index._1)))
-      .map(_.map(secureEnclaveApiClient.signWithEnclave(messageToSign, _)))
-      .map(Future.sequence(_))
-      .flatten
-
-    Try(Await.result(signaturesFromEnclaveFuture, timeoutDuration).flatten)
-      .getOrElse(Seq())
+    val keysFromEnclaveFuture = secureEnclaveApiClient.listPublicKeys()
+    val listOfKeys = Try(Await.result(keysFromEnclaveFuture, secureEnclaveRequestTimeout))
+      .map(managedKeys => indexedPublicKeys.filter(key_index => managedKeys.contains(key_index._1))).getOrElse(Seq())
+
+    val steps = listOfKeys.grouped(numOfParallelRequests)
+    val listOfSignatures = for (step <- steps) yield {
+      val listOfSignaturesFutures = step.map(secureEnclaveApiClient.signWithEnclave(messageToSign, _))
```

Comment:
:100: 

Let's add a comment to this method, I should have added it before, like this:
```
  /**
   * Get signatures from Secure Enclave for given public keys.
   * Ensures that no more than `numOfParallelRequests` are sent to Enclave at the same time.
   * If a batch fails to produce a response in `secureEnclaveRequestTimeout`, take what we have and continue.
   */
   ```